numpyro/contrib/funsor/discrete.py

```diff
@@ -97,27 +57,14 @@ def _sample_posterior(
             model_trace = trace(seed(model, rng_key)).get_trace(*args, **kwargs)
         first_available_dim = -_guess_max_plate_nesting(model_trace) - 1
 
-    with block(), enum(first_available_dim=first_available_dim):
-        with plate_to_enum_plate():
-            model_tr = packed_trace(model).get_trace(*args, **kwargs)
-
-    terms = terms_from_trace(model_tr)
-    # terms["log_factors"] = [log p(x) for each observed or latent sample site x]
-    # terms["log_measures"] = [log p(z) or other Dice factor
-    #                          for each latent sample site z]
-
-    with funsor.interpretations.lazy:
-        log_prob = funsor.sum_product.sum_product(
-            sum_op,
-            prod_op,
-            list(terms["log_factors"].values()) + list(terms["log_measures"].values()),
-            eliminate=terms["measure_vars"] | terms["plate_vars"],
-            plates=terms["plate_vars"],
-        )
-        log_prob = funsor.optimizer.apply_optimizer(log_prob)
+    with funsor.adjoint.AdjointTape() as tape:
+        with block(), enum(first_available_dim=first_available_dim):
+            log_prob, model_tr, log_measures = _enum_log_density(
+                model, args, kwargs, {}, sum_op, prod_op
+            )
 
     with approx:
-        approx_factors = funsor.adjoint.adjoint(sum_op, prod_op, log_prob)
+        approx_factors = tape.adjoint(sum_op, prod_op, log_prob)
```

Comment:
Does using the more functional interface `funsor.adjoint.adjoint` not work here?

Reply:
Sorry for the late response, @eb8680! Yes, for some reason there is an error happens at this line.
```
  File "/home/fehiepsi/numpyro/numpyro/contrib/funsor/discrete.py", line 120, in _sample_posterior
    approx_factors = funsor.adjoint.adjoint(sum_op, prod_op, log_prob)
  File "/home/fehiepsi/funsor/funsor/adjoint.py", line 135, in adjoint
    return tape.adjoint(sum_op, bin_op, root)
  File "/home/fehiepsi/funsor/funsor/adjoint.py", line 115, in adjoint
    in_adjs = adjoint_ops(fn, sum_op, bin_op, adjoint_values[output], *inputs)
  File "/home/fehiepsi/funsor/funsor/registry.py", line 106, in __call__
    return self[key](*args)
  File "/home/fehiepsi/funsor/funsor/registry.py", line 63, in __call__
    return self.partial_call(*args)(*args)
  File "/home/fehiepsi/funsor/funsor/adjoint.py", line 209, in adjoint_contract_generic
    assert len(terms) == 1 or len(terms) == 2
AssertionError
```